### ISSUE TITLE
[Feature] Force scripts update from an external addon

### ISSUE THREAD

**Issue by pavolbiely** [violentmonkey/violentmonkey#1777]
**Motivation**
I want to force script updates with my own extension in Violentmonkey.

**Proposed Solution**
Implement [runtime.onMessageExternal](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessageExternal) event which will listen to calls which would force scripts updates.

Add new Violentmonkey event listener like this:
```
browser.runtime.onMessageExternal.addListener((message, sender) => {
  if (message == "force-scripts-update") {
    // process script updates
  }
});
```

I will call it with my extension like this:
```
browser.runtime.sendMessage("violentmonkey@addonId", "force-scripts-update");
```

**Use Cases**
My scripts are sometimes changing more often than once an hour or they need to be updated on-demand.

**Comment by tophf** [comment]
I don't think we should implement such a niche feature because we will have to manually list the ids of the extensions allowed to communicate in manifest.json (there's no API to customize them dynamically).

For hot-reloading of a frequently updated code while developing we already have [tracking in in the installer tab](https://violentmonkey.github.io/posts/how-to-edit-scripts-with-your-favorite-editor/). In the future we may implement other means of tracking that don't require having an open tab, see #430, #1089.

WDYT, @gera2ld? What do you want to do with the [unused remnants of the plugin API](https://github.com/violentmonkey/violentmonkey/blob/master/src/background/plugin/index.js)? I would rather just remove them.

**Comment by pavolbiely** [comment]
By default, an extension can receive messages from any other extension. If externally_connectable in manifest.json is not specified, all extensions can communicate with each other but not with web pages. So manual IDs list of allowed extension is not required.

**Comment by tophf** [comment]
Ah, then technically we can implement it and add a list of allowed extension ids in settings.
The other and bigger questions remain though.

**Comment by tophf** [comment]
Actually, I don't like our dead plugin API because it doesn't seem useful enough to have anyone use it while requiring us to maintain it and provide backward compatibility, **so a possible solution would be to expose our entire internal API** to the user-allowed extensions. We would absolve ourselves from maintaining backward compatibility and the need to maintain the API. Since we don't change our internal API often it's quite stable as-is. In the future, if we ever see more demand for this we can add customizable permissions like github does for external apps i.e. allow only some commands or full control. We would show a warning banner in the popup and dashboard for the latter.

**Comment by gera2ld** [comment]
The dead plugin API was designed for exactly this purpose because people asked to use file system as the script storage but I didn't want to connect to the file system by myself.

> a possible solution would be to expose our entire internal API

I think it would be better to just expose few high level APIs doing limited stuff to lower the risk of being broken/hacked by other extensions.

**Comment by tophf** [comment]
I don't see any way how it can be hacked. The user intentionally allows just this extension to control Violentmonkey and the user installs the extension.

**Comment by tophf** [comment]
Thing is, how should we design this public API and why? The current plugin API doesn't make sense as its only consumer is our sync and this is weird. It should just use the internal commands directly, which will finally allow us to remove the dead remnants.

**Comment by tophf** [comment]
For example, take the current `update` method in plugin API. It vaguely and misleadingly implies it'll check the script for an update while actually it installs the script. Or the `list` method which implies it'll list the names of the scripts but actually returns the entire data array. Or the `get` method which is so vague that I can't even say what it does by its name.

Designing a public API is not easy. Having two API systems with different names for the same methods makes it hard to maintain. I suggest we remove the weird public plugin API completely and switch our sync to use our own API regardless of whether we expose it or not.

**Comment by gera2ld** [comment]
> how should we design this public API and why?

Sorry it's been a long time, I might be wrong about the purpose. 
IIRC there could be another plugin to provide the same interface and allow the sync service to sync the scripts to a different backend (e.g. file system). So the plugin is more like a way to define the storage.
The update/list/get/remove here is similar to REST APIs of a storage service.

I agree to remove it since it's not used after all.

**Comment by tophf** [comment]
So about exposing the internal API, what about `CheckUpdate(id)` command for the use case in this issue (allowed for a specified extension ids)? We might expose more in the future when a use case is presented to us.

**Comment by gera2ld** [comment]
> what about `CheckUpdate(id)` command for the use case in this issue (allowed for a specified extension ids)?

I'm ok with that.

BTW I don't think we need to worry about backward compatibility. This feature is definitely not for average users and likely won't be used by many users in the foreseeable future, so it should not be a problem for external plugins to adapt to API changes.
